numpyro/infer/elbo.py

```diff
@@ -280,3 +281,266 @@ def single_particle_elbo(rng_key):
         renyi_elbo = avg_log_exp / (1.0 - self.alpha)
         weighted_elbo = jnp.dot(stop_gradient(weights), elbos) / self.num_particles
         return -(stop_gradient(renyi_elbo - weighted_elbo) + weighted_elbo)
+
+
+def _get_plate_stacks(trace):
+    """
+    This builds a dict mapping site name to a set of plate stacks. Each
+    plate stack is a list of :class:`CondIndepStackFrame`s corresponding to
+    an :class:`plate`. This information is used by :class:`Trace_ELBO` and
+    :class:`TraceGraph_ELBO`.
+    """
+    return {
+        name: [f for f in node["cond_indep_stack"]]
+        for name, node in trace.items()
+        if node["type"] == "sample"
+    }
+
+
+class MultiFrameTensor(dict):
```

Comment:
can you please confirm which chunks of code were borrowed verbatim apart from torch/jnp edits?

Reply:
yes, it is just the same here: only torch -> jnp